libcrux-ml-kem/tests/acvp.rs

```diff
@@ -0,0 +1,345 @@
+}
+
+#[derive(Deserialize)]
+#[allow(non_snake_case, dead_code)]
+struct EncapDecapPromptTestGroup {
+    tgId: usize,
+    testType: String,
+    parameterSet: String,
+    #[serde(flatten)]
+    tests: EncapDecapTests,
+}
+
+#[allow(non_snake_case, dead_code)]
+#[derive(Deserialize)]
+#[serde(tag = "function")]
+enum EncapDecapTests {
+    #[serde(rename(deserialize = "encapsulation"))]
+    EncapTests { tests: Vec<EncapPrompt> },
+    #[serde(rename(deserialize = "decapsulation"))]
+    DecapTests {
+        #[serde(with = "hex::serde")]
+        dk: Vec<u8>,
+        tests: Vec<DecapPrompt>,
+    },
+}
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+#[allow(non_snake_case, dead_code)]
+enum EncapDecapResult {
+    EncapResult {
+        tcId: usize,
+        #[serde(with = "hex::serde")]
+        c: Vec<u8>,
+        #[serde(with = "hex::serde")]
+        k: Vec<u8>,
+    },
+    DecapResult {
+        tcId: usize,
+        #[serde(with = "hex::serde")]
+        k: Vec<u8>,
+    },
+}
+
+#[derive(Deserialize)]
+#[allow(non_snake_case, dead_code)]
+struct ResultEncapDecapTestGroup {
+    tgId: usize,
+    tests: Vec<EncapDecapResult>,
+}
+
+#[test]
+fn encap_decap() {
+    use libcrux_ml_kem::*;
+
+    let prompts: Prompts<EncapDecapPromptTestGroup> = read("encap-decap", "prompt.json");
+    assert!(prompts.algorithm == "ML-KEM");
+    assert!(prompts.revision == "FIPS203");
+
+    let results: Results<ResultEncapDecapTestGroup> = read("encap-decap", "expectedResults.json");
+    assert!(results.algorithm == "ML-KEM");
+    assert!(results.revision == "FIPS203");
+
+    for kat in prompts.testGroups {
+        let parameter_set = kat.parameterSet;
+        assert!(
+            parameter_set == "ML-KEM-512"
+                || parameter_set == "ML-KEM-768"
+                || parameter_set == "ML-KEM-1024"
+        );
+        eprintln!("{parameter_set}");
+
+        match kat.tests {
+            EncapDecapTests::EncapTests { tests } => {
+                for test in tests {
+                    let expected_result = results
+                        .testGroups
+                        .iter()
+                        .find(|tg| tg.tgId == kat.tgId)
+                        .unwrap()
+                        .tests
+                        .iter()
+                        .find(|t| {
+                            if let EncapDecapResult::EncapResult { tcId, .. } = t {
+                                *tcId == test.tcId
+                            } else {
+                                false
+                            }
+                        })
+                        .unwrap();
+
+                    if let EncapDecapResult::EncapResult { c, k, .. } = expected_result {
+                        let ek = test.ek;
+                        let randomness = test.m;
+                        match parameter_set.as_str() {
+                            #[cfg(feature = "mlkem512")]
+                            "ML-KEM-512" => {
```

Comment:
Nothing for here. But we should really add that dynamic API on top.

Reply:
Can you elaborate? Do you mean provided randomness vs internally generated?